ubirch/ubirch_protocol_test.go

```diff
@@ -620,67 +1101,137 @@ func TestDecode(t *testing.T) {
+	r := rand.Reader                             //the RNG under test
+	nBytes := 256                                //amount of random bytes to be tested
+	nBits := nBytes * 8                          //amount of bits in the random data
+	randomBytesUnderTest := make([]byte, nBytes) //the random data to be tested for randomness
+	_, err := io.ReadFull(r, randomBytesUnderTest)
+	requirer.NoError(err, "Generating random bytes failed: %v", err)
+
+	//calculate the frequency of ones and zeros in the random data
 	s := 0
-	for i := 0; i < n; i++ {
+	for i := 0; i < nBytes; i++ {
 		// get number of one bits (population count)
-		ones := bits.OnesCount8(randomNumberUnderTest[i])
+		ones := bits.OnesCount8(randomBytesUnderTest[i])
 		// count +1 for every one bit and -1 for every zero bit
 		s += (2 * ones) - 8
 	}
-	log.Printf("s: %v", s)
-
 	// calculate the test statistic
-	s_obs := math.Abs(float64(s)) / math.Sqrt(float64(n))
+	sObs := math.Abs(float64(s)) / math.Sqrt(float64(nBits))
+	pValueCalc := math.Erfc(sObs / math.Sqrt2)
+
+	//Debug info
+	//log.Printf("s: %v", s)
+	//log.Printf("Calculated pValue: %v", pValueCalc)
+	// for i := 0; i < nBytes-8; {
+	// 	log.Printf("%08b%08b%08b%08b%08b%08b%08b%08b\n",
+	// 		randomBytesUnderTest[i], randomBytesUnderTest[i+1],
+	// 		randomBytesUnderTest[i+2], randomBytesUnderTest[i+3],
+	// 		randomBytesUnderTest[i+4], randomBytesUnderTest[i+5],
+	// 		randomBytesUnderTest[i+6], randomBytesUnderTest[i+7])
+	// 	i += 8
+	// }
+
+	//Decision Rule: If the computed P-value is pValueCalc < pValueLimit, then conclude that the sequence is non-random.
+	//Otherwise, conclude that the sequence is random.
+	requirer.Greater(pValueCalc, pValueLimit, "Random data did not pass Frequency (Monobit) test. (pValueCalc was smaller than pValueLimit)\n"+
+		"This means with 99 %% confidence that something is wrong, but in 1 %% of tests there is a false positive.\n"+
+		"Random data was :\n%v", randomBytesUnderTest)
+}
 
-	pValue := math.Erfc(s_obs / math.Sqrt2)
-	log.Printf("pValue: %v", pValue)
+//TestECDSASignatureChanges tests if the signature of ECDSA changes for the same input using the top level protocol struct.
+//If it does not, there is likely a problem with the random number or nonce ('k') generation
+//which will allow attackers to calculate the private key from signatures.
+//Since only a 'small' number of signatures is checked this will most likely detect only 'total' failures
+//in k/nonce generation. See also  https://en.wikipedia.org/wiki/Elliptic_Curve_Digital_Signature_Algorithm#Security
+//For testing, 'signed' type UPPs are used. Both "new context for each UPP" and "one context across all UPPs" cases are tested
+func TestECDSASignatureChanges(t *testing.T) {
+	const nrOfSigsToCheck = 1000 //effective number of tests is two time this number as both "new context for each test" and "consistent context" cases are tested
+	userDataBytes := []byte("Hello World!")
```

Comment:
Maybe better to use default input data global constant here for readability?

Reply:
fixed